be/src/exec/stream/aggregate/stream_aggregator.h

```diff
@@ -0,0 +1,164 @@
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <algorithm>
+#include <utility>
+
+#include "exec/stream/aggregate/agg_state_data.h"
+#include "exec/stream/state/mem_state_table.h"
+#include "exec/stream/state/state_table.h"
+#include "exec/vectorized/aggregator.h"
+#include "runtime/mem_pool.h"
+#include "runtime/runtime_state.h"
+#include "runtime/types.h"
+
+namespace starrocks::stream {
+
+/**
+ * StreamAggregator is used for stream materialized view, it differs with Aggregator:
+ * stream aggregator's state is not kept into mem_pool, rather in the StateCache which can be changed adaptively;
+ * stream aggregator's processing is by row rather than by chunk which is better in stream materialized view;
+ * stream aggregator handles input's chunk as below:
+ *  - step1: compute the agg state:
+ *      - iterate each row of input chunk,
+ *      - 1.1 if its state is in StateCache, get the state's address
+ *      - 1.2 query the aggregate StateTable:
+ *          - 1.2.1 if it is in StateTable, deserialized the result and store in the StateCache, then redo 1.1
+ *          - 1.2.2 insert a new state into StateCache, and get the state's address
+ *  - step2: update the intermediate agg state:
+ *      - iterate each row of input chunk, call aggregate functions' update functions to accumulate the state.
+ *  - step3: update StateTable
+ *      - iterate incremental input rows and write back into the StateTable
+ *  - step4: output the incremental results
+ *      - iterate incremental input rows and output to the next operator
+ */
+class StreamAggregator final : public Aggregator {
+public:
+    StreamAggregator(std::shared_ptr<AggregatorParams>&& params);
+
+    static constexpr StreamRowOp INSERT_OP = StreamRowOp::INSERT;
+    static constexpr StreamRowOp DELETE_OP = StreamRowOp::DELETE;
+    static constexpr StreamRowOp UPDATE_BEFORE_OP = StreamRowOp::UPDATE_BEFORE;
+    static constexpr StreamRowOp UPDATE_AFTER_OP = StreamRowOp::UPDATE_AFTER;
+
+    ~StreamAggregator() override {
+        if (_state != nullptr) {
+            close(_state);
+        }
+    }
+
+    Status open(RuntimeState* state) { return Aggregator::open(state); }
+
+    Status prepare(RuntimeState* state, ObjectPool* pool, RuntimeProfile* runtime_profile,
+                   MemTracker* mem_tracker) override {
+        RETURN_IF_ERROR(Aggregator::prepare(state, pool, runtime_profile, mem_tracker));
+        RETURN_IF_ERROR(_prepare_state_tables(state));
+        return Status::OK();
+    }
+
+    // Process input's chunks util `Epoch` chunk is received.
+    Status process_chunk(vectorized::StreamChunk* chunk);
+
+    Status output_changes(int32_t chunk_size, vectorized::StreamChunkPtr* result_chunk);
+
+    // Called when need to generate incremental outputs and Output agg_states for the next batch.
+    Status output_changes(int32_t chunk_size, vectorized::StreamChunkPtr* result_chunk,
```

Comment:
What's the difference of these two `output_changes` ?

Reply:
the latter one is to used for testing.